src/store/assets/actions.ts

```diff
@@ -1,47 +1,149 @@
+        balance: ZeroStringValue,
+        decimals: +asset.externalDecimals,
+      };
+      return buffer;
+    }, {});
+  } else {
+    registeredAssets = { ...state.registeredAssets };
+  }
+
+  const networkAssetsWithBalance = await Promise.all(
+    Object.entries(registeredAssets).map(async ([soraAddress, assetData]) => {
+      const accountAsset = { ...assetData };
+      try {
+        if (!accountAsset.address) {
+          accountAsset.address = await rootDispatch.web3.getEvmTokenAddressByAssetId(soraAddress);
+        }
+        if (accountAsset.address) {
+          const { value, decimals } = await ethersUtil.getAccountAssetBalance(accountAddress, accountAsset.address);
+
+          accountAsset.balance = value;
+          accountAsset.decimals = decimals;
+
+          // update evmBalance
+          if (ethersUtil.isNativeEvmTokenAddress(accountAsset.address)) {
+            rootCommit.web3.setEvmBalance(value);
+          }
+        }
+      } catch (error) {
+        console.error(error);
+      }
+
+      return {
+        [soraAddress]: accountAsset,
+      };
+    })
+  );
+
+  return networkAssetsWithBalance;
+}
+
+async function getEvmRegisteredAssets(context: ActionContext<any, any>): Promise<Record<string, EvmAccountAsset>[]> {
+  const { rootState, rootCommit } = assetsActionContext(context);
+
+  const accountAddress = rootState.web3.evmAddress;
+  const evmNetworkId = rootState.web3.evmNetworkSelected;
+  const networkAssets = await evmBridgeApi.getNetworkAssets(evmNetworkId as number);
+
+  const networkAssetsWithBalance = await Promise.all(
+    Object.entries(networkAssets).map(async ([soraAddress, assetData]) => {
+      const accountAsset = {
+        address: assetData.evmAddress as string,
+        balance: ZeroStringValue,
+        contract: assetData.contract,
+        decimals: 18,
+      };
+
+      try {
+        const { value, decimals } = await ethersUtil.getAccountAssetBalance(accountAddress, accountAsset.address);
+
+        accountAsset.balance = value;
+        accountAsset.decimals = decimals;
+
+        // update evmBalance
+        if (ethersUtil.isNativeEvmTokenAddress(accountAsset.address)) {
+          rootCommit.web3.setEvmBalance(value);
+        }
+      } catch (error) {
+        console.error(error);
+      }
+
+      return {
+        [soraAddress]: accountAsset,
+      };
+    })
+  );
+
+  return networkAssetsWithBalance;
+}
 
 const actions = defineActions({
-  async updateRegisteredAssets(context, reset?: boolean): Promise<void> {
-    const { state, commit, rootDispatch } = assetsActionContext(context);
+  // for common usage
+  async updateRegisteredAssets(context, force?: boolean): Promise<void> {
+    const { state, commit, dispatch, rootState } = assetsActionContext(context);
 
-    if (state.registeredAssetsFetching) return;
+    if (state.registeredAssetsFetching) {
+      if (force) {
+        await delay(250);
+        return await dispatch.updateRegisteredAssets(force);
```

Comment:
Might it cause indefinite loop in some cases?

Reply:
This code awaits the execution of the previous `updateRegisteredAssets` request - so it could awaits at max 60 seconds - this is our ws request connection timeout